task_processing/plugins/kubernetes/kubernetes_pod_executor.py

```diff
@@ -107,21 +114,175 @@ def _pod_event_watch_loop(self) -> None:
+                f"Got a MODIFIED event for {pod_name} for unhandled phase: "
+                f"{pod.status.phase} - ignoring."
+            )
+            return
+
+        if (
+            pod.status.phase == "Succeeded"
+            and task_metadata.task_state is not KubernetesTaskState.TASK_FINISHED
+        ):
+            logger.info(
+                f"Removing {pod_name} from state and emitting 'finished' event.",
+            )
+            self.task_metadata = self.task_metadata.discard(pod_name)
+            self.event_queue.put(
+                task_event(
+                    task_id=pod_name,
+                    terminal=True,
+                    success=True,
+                    timestamp=time.time(),
+                    raw=event["raw_object"],
+                    task_config=task_metadata.task_config,
+                    platform_type="finished"
+                )
+            )
+            return
+
+        elif (
+            pod.status.phase == "Failed"
+            and task_metadata.task_state is not KubernetesTaskState.TASK_FAILED
+        ):
+            logger.info(f"Removing {pod_name} from state and emitting 'failed' event.")
+            self.task_metadata = self.task_metadata.discard(pod_name)
+            self.event_queue.put(
+                task_event(
+                    task_id=pod_name,
+                    terminal=True,
+                    success=False,
+                    timestamp=time.time(),
+                    raw=event["raw_object"],
+                    task_config=task_metadata.task_config,
+                    platform_type="FAILED"
+                )
+            )
+            return
+
+        elif (
+            pod.status.phase == "Running"
+            and task_metadata.task_state is not KubernetesTaskState.TASK_RUNNING
+        ):
+            logger.info(f"Successfully launched {pod_name}, emitting 'running' event.")
+            self.task_metadata = self.task_metadata.set(
+                pod_name,
+                task_metadata.set(
+                    node_name=pod.status.host_ip,
+                    task_state=KubernetesTaskState.TASK_RUNNING,
+                    task_state_history=task_metadata.task_state_history.append(
+                        (KubernetesTaskState.TASK_RUNNING, time.time()),
+                    )
+                )
+            )
+            self.event_queue.put(
+                task_event(
+                    task_id=pod_name,
+                    terminal=False,
+                    timestamp=time.time(),
+                    raw=event["raw_object"],
+                    task_config=task_metadata.task_config,
+                    platform_type="running"
+                )
+            )
+            return
+
+        # XXX: figure out how to handle this correctly (and when this actually
+        # happens - we were unable to cajole k8s into giving us an event with an Unknown
+        # phase)
+        elif pod.status.phase == "Unknown":
```

Comment:
Should we be checking if we'd already set the task_metadata to Unknown as well, just in case?

Reply:
yea, won't hurt!